### PR TITLE
Split serving assets into two different in-repo addons

### PR DESCRIPTION
**Pull request by kratiahuja**
Implementation of [RFC # 80 of ember-cli](https://github.com/ember-cli/rfcs/pull/80). This PR primarily does the following:
- [x] Introduces `ember-cli:broccoli:watcher` addon that is responsible for setting the response headers for every asset request being served.
- [x] Introduces `ember-cli:broccoli:serve-files` addon that is responsible for serving the assets
- [x] Removes `serve-files` addon since the above two addon do the same work
- [x] Release and update `broccoli-middleware` version after [PR](https://github.com/ember-cli/broccoli-middleware/pull/21) is merged.
- [x] Fix broken tests here after the `broccoli-middleware` version is updated.
- [x] Update `ember-cli-inject-livereload` from `serve-files` to `ember-cli:broccoli:serve-files` [here](https://github.com/ember-cli/ember-cli-inject-live-reload/pull/39)

**Note**: `broccoli-middleware` output format is now changed but to maintain backward compatibility it still has the original one middleware that does both setting response headers and serving assets.

TODO:
- [ ] Release a major version of `ember-cli-inject-live-reload` that is compatible with the first release of ember-cli containing this PR.
- [ ] Update change log?

### CODE REVIEW

File: lib/tasks/server/middleware/ember-cli:broccoli:serve-files/index.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
+'use strict';
+
+var cleanBaseURL = require('clean-base-url');
+var logger       = require('heimdalljs-logger')('ember-cli:broccoli:serve-files');
+
+/**
+ * This addon is used to set the response headers for the assets that will be
+ * served by the next middleware. It waits for the watcher promise to resolve before
+ * setting the response headers.
```
Comment (hjdivad) [inline lib/tasks/server/middleware/ember-cli:broccoli:serve-files/index.js]: Is this comment right?  it looks like response headers are actually set [in the watcher](https://github.com/ember-cli/broccoli-middleware/blob/aac8215f5184b226c51e964313c5c49460cf0da1/lib/watcher-middleware.js#L32)

File: lib/tasks/server/middleware/ember-cli:broccoli:watcher/index.js [hunk 2]
```diff
@@ -1,25 +1,25 @@
...
 
 var cleanBaseURL = require('clean-base-url');
-var logger       = require('heimdalljs-logger')('ember-cli:serve-files');
+var logger       = require('heimdalljs-logger')('ember-cli:broccoli:watcher');
 
 /**
- * This addon is used to serve the app (`index.html`) and
- * assets (`*.js`, `*.css`, ...) at the `baseURL` prefix.
+ * This addon is used to serve the requested assets. It runs after ember-cli:broccoli:watcher
+ * addon
```
Comment (hjdivad) [inline lib/tasks/server/middleware/ember-cli:broccoli:watcher/index.js]: Is this comment right?  Isn't this the addon that installs the `ember-cli:broccoli:watcher` middleware, not one that runs after it?

File: lib/tasks/server/middleware/ember-cli:broccoli:serve-files/index.js (hunk 1 above)
Comment (kratiahuja) [inline lib/tasks/server/middleware/ember-cli:broccoli:serve-files/index.js]: Sorry my bad, I messed up the comments in both. Will fix it.

File: lib/tasks/server/middleware/ember-cli:broccoli:watcher/index.js (hunk 2 above)
Comment (kratiahuja) [inline lib/tasks/server/middleware/ember-cli:broccoli:watcher/index.js]: Sorry my bad, I messed up the comments in both. Will fix it.

**Comment by kratiahuja** [comment]
@stefanpenner @hjdivad @rwjblue : This is ready for ready for review. 

I am not sure why appveyor tests are failing saying it can't create the new files. Could someone please help me debug appveyor issue?

**Comment by stefanpenner** [comment]
> We also need to update the changelog. I am not sure if that is handrolled before the release?

You can skip that, we generate the changelog as part of our release.

**Comment by stefanpenner** [comment]
> I am not sure why appveyor tests are failing saying it can't create the new files. Could someone please help me debug appveyor issue?

It seems windows filesystem does snot support `:` in a file name, I believe changing the file names to use `-` instead maya ddress the issue.

**Comment by stefanpenner** [comment]
> NOTE: Since ember-cli-inject-live-reload depends on the old middleware, it will need a major version bump I believe.

ping me tomorrow and I can release :)

**Comment by stefanpenner** [comment]
looking good, but I'm going to give it another review when I am less sleepy :P

**Comment by kratiahuja** [comment]
@stefanpenner Sounds perfect! Thank you! :)

File: lib/models/project.js [hunk 3]
```diff
@@ -349,7 +349,8 @@ class Project {
       path.join(internalMiddlewarePath, 'testem-url-rewriter'),
       path.join(internalMiddlewarePath, 'tests-server'),
       path.join(internalMiddlewarePath, 'history-support'),
-      path.join(internalMiddlewarePath, 'serve-files'),
+      path.join(internalMiddlewarePath, 'ember-cli:broccoli:watcher'),
+      path.join(internalMiddlewarePath, 'ember-cli:broccoli:serve-files'),
```
Comment (stefanpenner) [inline lib/models/project.js]: i believe windows will be unhappy with `:` in the filenames.

**Comment by stefanpenner** [comment]
> NOTE: Since ember-cli-inject-live-reload depends on the old middleware name, therefore it will need a major version bump I believe. If we need to make it backward compatible, we will need to rename ember-cli:broccoli:serve-files to be called serve-files-middleware.

Would releasing a new major version of ember-cli-inject-live-reload which uses the new names work?

**Comment by rwjblue** [comment]
We do not need a major bump of ember-cli-inject-live-reload, v1.6.0+ will work properly before and after this PR.

**Comment by hjdivad** [comment]
@rwjblue cool, I didn't realise the addon dep graph setup was this resilient.

**Comment by hjdivad** [comment]
> NOTE: Since ember-cli-inject-live-reload depends on the old middleware name, therefore it will need a major version bump I believe. If we need to make it backward compatible, we will need to rename ember-cli:broccoli:serve-files to be called serve-files-middleware.

@kratiahuja seems like you and @rwjblue already sorted this out in https://github.com/ember-cli/ember-cli-inject-live-reload/pull/39

👍

**Comment by kratiahuja** [comment]
@hjdivad yup, I wasn't aware of that magic 😄  . Will update this PR to remove `:` and have `-`. Also since the file name will now look pretty long, I am going to drop `ember-cli`.

File: lib/models/project.js (hunk 3 above)
Comment (kratiahuja) [inline lib/models/project.js]: Fixed with replacing it with `-` and dropping `ember-cli` since the name was turning too long.

**Comment by kratiahuja** [comment]
appveyor is now failing on unrelated issue.

**Comment by rwjblue** [comment]
LETS DO IT!

@homu r+

**Comment by kratiahuja** [comment]
Yayy! Thank you @stefanpenner @rwjblue @hjdivad for the feedback and helping me through this!

**Pull request merged**